daemon/containerd/image_exporter.go

```diff
@@ -115,6 +129,84 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
 		}
 
 		i.LogImageEvent(target.Digest.String(), target.Digest.String(), events.ActionSave)
+		return nil
+	}
+
+	exportRepository := func(ctx context.Context, ref reference.Named) error {
+		imgs, err := i.getAllImagesWithRepository(ctx, ref)
+		if err != nil {
+			return errdefs.System(fmt.Errorf("failed to list all images from repository %s: %w", ref.Name(), err))
+		}
+
+		if len(imgs) == 0 {
+			return images.ErrImageDoesNotExist{Ref: ref}
+		}
+
+		for _, img := range imgs {
+			ref, err := reference.ParseNamed(img.Name)
+
+			if err != nil {
+				log.G(ctx).WithFields(log.Fields{
+					"image": img.Name,
+					"error": err,
+				}).Warn("couldn't parse image name as a valid named reference")
+				continue
+			}
+
+			if err := exportImage(ctx, img.Target, ref); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	for _, name := range names {
+		target, resolveErr := i.resolveDescriptor(ctx, name)
+
+		// Check if the requested name is a truncated digest of the resolved descriptor.
+		// If yes, that means that the user specified a specific image ID so
+		// it's not referencing a repository.
+		specificDigestResolved := false
+		if resolveErr == nil {
+			nameWithoutDigestAlgorithm := strings.TrimPrefix(name, target.Digest.Algorithm().String()+":")
+			specificDigestResolved = strings.HasPrefix(target.Digest.Encoded(), nameWithoutDigestAlgorithm)
+		}
+
+		log.G(ctx).WithFields(log.Fields{
+			"name":                   name,
+			"resolveErr":             resolveErr,
+			"specificDigestResolved": specificDigestResolved,
+		}).Debug("export requested")
+
+		ref, refErr := reference.ParseNormalizedNamed(name)
+
+		if resolveErr != nil || !specificDigestResolved {
+			// Name didn't resolve to anything, or name wasn't explicitly referencing a digest
+			if refErr == nil && reference.IsNameOnly(ref) {
+				// Reference is valid, but doesn't include a specific tag.
+				// Export all images with the same repository.
+				if err := exportRepository(ctx, ref); err != nil {
+					return err
+				}
+				continue
+			}
+		}
+
+		if resolveErr != nil {
+			return resolveErr
+		}
+		if refErr != nil {
+			return refErr
+		}
```

Comment:
This is gross and ugly, but I didn't had an idea how to make it better while maintaining the overall behavior 🙈 

Please let me know if you have suggestions here.

Reply:
I think it's good to implement these oddities, but we should keep a list, and create tracking issues. Some of those (from the top of my head);


- API: `image` (without `:tag`) on `push` and `pull` meaning "all tags".
    - I _THINK_ there's a tracking issue for that, but we should deprecate that (possibly as breaking change or "versioned"), and require an `?all-tags` option in the API.
    - For `pull` we could even consider deprecating _entirely_, as this involves querying (scraping) the registry to get a list of all tags
    - Instead, we could _require_ a list of tags to pull (or just one "request per tag") (TBD)
- For `image save`: similar;
    - Make sure we have some endpoint that allows returning a list of tags for images with the same `name`
    - That endpoint can be used by the client to query the list
    - After which the client can send the list of `image:tag`'s to save
    - (We can provide a utility function for this in the sdk / client to handle all this)

Reply:
@thaJeztah maybe open an issue, this comment will be lost

Reply:
> Make sure we have some endpoint that allows returning a list of tags for images with the same name

Wouldn't `docker images --filter "reference=<repository>:*"` be enough here?

Reply:
> Although, it's slightly less atomic, as theoretically the image could be deleted/some other image be added in between.

Yeah, but also not sure if the current daemon-side code _actually_ protects against that case; we're also iterating over names, and don't "lock the world" during the export, so it's gonna be racy in either case.  Time between getting that list and using it will be shorter, but we can't provide any certainty as well (just "slightly less likely to be hit")